File: logic/classes/Expert.py
import os
from classes.Fact import Fact
from classes.Rule import Rule
from classes.ErrorHandler import error_handler
from validation.validate import validate_line, validate_known_data


class Expert:
	def __init__(self, file_path, ui_flag):
		self.facts_flag = False
		self.known_facts = []
		self.queries = []
		self.rules = []
		self.facts = {}
		self.result = []

		self.path = file_path
		self.ui = ui_flag

	@staticmethod
	def get_known_data(line):
		data = []
		for char in line:
			data.append(char)
		return data

	@staticmethod
	def clean_up_line(line):
		if '#' in line:
			line = line.split('#')[0]
		for char in [' ', '\t', '\n', '\f', '\v']:
			if char in line:
				line = line.replace(char, '')

		return line

	@staticmethod
	def manage_implication(rule, rules_list):
		if '<=>' in rule:
			split = rule.split('<=>')
			new_rule = split[1] + '=>' + split[0]
			rules_list.append(new_rule)
			rule = rule.replace('<=>', '=>')
		rules_list.append(rule)
		return rules_list

	@staticmethod
	def convert_rule(rule):
		return '!({})=>!({})'.format(rule.implication, rule.condition)

	def validate_data(self):
		errors = validate_known_data(self.known_facts, self.queries, self.facts_flag)
		if errors or error_handler.errors:
			if errors:
				error_handler.errors['global'] = errors
			error_handler.throw_errors()

	def log_result(self):
		with open(os.path.join('..', 'result', 'log'), 'w') as f:
			for result in self.result:
				f.write(result + '\n')
				print result

	def validate_file(self):
		if self.ui == True:
			return True
		try:
			with open(self.path, 'r') as f:
				return True
		except IOError:
			error_handler.throw_error('global', 'File "{}" does not appear to exist.'.format(self.path))

	def parse_file(self):
		line_counter = 0

		if self.ui:
			split = self.path.split('\n')
			for line in split:
				line_counter += 1
				self.manage_line(line, line_counter)
		else:
			with open(self.path, 'r') as f:
				for line in f:
					line_counter += 1
					self.manage_line(line, line_counter)

	def manage_line(self, line, i):
		line = self.clean_up_line(line)
		if not line:
			return

		errors = validate_line(line)
		if errors:
			error_handler.errors[i] = errors
		first_char = line[0]
		if first_char == '=' or first_char == '?':
			if first_char == '=':
				self.facts_flag = True
				self.known_facts = self.get_known_data(line[1:])
			if first_char == '?':
				self.queries = self.get_known_data(line[1:])
		else:
			self.rules.append(line)

	def manage_facts(self):
		for fact in self.queries:
			self.facts[fact] = Fact(fact, self)
		for fact in self.known_facts:
			self.facts[fact] = Fact(fact, self)
			self.facts[fact].state = 'true'
			self.facts[fact].resolved = True

	def manage_rules(self):
		temp_rules_list = []
		for rule in self.rules:
			temp_rules_list = self.manage_implication(rule, temp_rules_list)

		self.rules = []
		for rule in temp_rules_list:
			self.rules.append(Rule(rule, self))

		for rule in self.rules:
			Rule(self.convert_rule(rule), self)

	def manage_queries(self):
		for query in self.queries:
			result = self.facts[query].resolve()
			if result == 'none':
				result = 'false (default)'
			self.result.append('{} is {}'.format(query, result))

File: logic/classes/Expert.js
const { validateLine, validateKnownData } = require('../validation/validate');
const { errorHandler } = require('./ErrorHandler');
const { Rule } = require('./ErrorHandler');
const { Fact } = require('./ErrorHandler');
const fs = require('fs');


class Expert {
    constructor(filePath) {
        this.factsFlag = false;
		this.knownFacts = [];
		this.queries = [];
		this.rules = [];
		this.facts = {};
		this.result = [];
		this.path = filePath;
    }
	static getKnownData(line) {
        return line.split();
    }
	static cleanUpLine(line) {
        if (line.includes('#')) {
            line = line.split('#')[0];
        }
		for (const char of [' ', '\t', '\n', '\f', '\v']) {
            if (line.includes(char)) {
                line = line.replace(char, '');
            }
        }
        return line
    }
	static manageImplication(rule, rulesList) {
        if (rule.includes('<=>')) {
            const split = rule.split('<=>');
			const newRule = split[1] + '=>' + split[0];
			rulesList.push(newRule);
			rule = rule.replace('<=>', '=>');
        }
		rulesList.push(rule);
		return rulesList
    }
	static convertRule(rule) {
        return `!(${rule.implication})=>!(${rule.condition})`;
    }
	validateData() {
        const errors = validateKnownData(this.knownFacts, this.queries, this.factsFlag);
		if (errors || errorHandler.errors) {
		    if (errors) {
                errorHandler.errors['global'] = errors;
            }
			errorHandler.throwErrors();
        }
    }
	logResult() {
        for (const result of this.result) {
            console.log(result);
        }
    }
	validateFile() {
        if (this.ui === true) {
            return true;
        }
        if (!fs.existsSync(this.path)) {
            errorHandler.throwError('global', `File ${this.path} does not appear to exist.`)
        }
    }
	parseFile() {
        let i = 0;
        const data = fs.readFileSync(this.path);
        for (const line of data.split('\n')) {
            i++;
            this.manageLine(line, i)
        }
    }
	manageLine(line, i) {
        line = this.cleanUpLine(line);
		if (!line) { // todo  check
		    return;
        }
		const errors = validateLine(line);
		if (errors) {
            errors.errors[i] = errors;
        }
        const firstChar = line.charAt(0);
        if (firstChar === '=' || firstChar === '?') {
            if (firstChar === '=') {
                this.factsFlag = true;
                this.knownFacts = this.constructor.getKnownData(line.slice(1));
            }
            if (firstChar === '?') {
                this.queries = this.constructor.getKnownData(line.slice(1))
            }
        } else {
            this.rules.push(line)
        }
    }
	manageFacts() {
        for (const fact of this.queries) {
            this.facts[fact] = new Fact(fact, this);
        }
		for (const fact of this.knownFacts) {
		    this.facts[fact] = new Fact(fact, this);
			this.facts[fact].state = 'true';
			this.facts[fact].resolved = true;
        }
    }
	manageRules() {
        let tempRulesList = [];
		for (const rule of this.rules) {
		    tempRulesList = this.constructor.manageImplication(rule, tempRulesList)
		}
		this.rules = [];
		for (const rule of tempRulesList) {
		    this.rules.push(new Rule(rule, this));
		}
		for (const rule of this.rules) {
		    new Rule(this.constructor.convertRule(rule), this); // TODO EXTRA WHAAAAAAT????
        }
    }
	manageQueries() {
        for (const query of this.queries) {
            let result = this.facts[query].resolve();
			if (result === 'none') {
			    result = 'false (default)';
            }
			this.result.push(`${query} is ${result}`);
        }
    }
}

module.exports = {
    Expert
};

